### PR TITLE
[SPARK-21743][SQL][followup] free aggregate map when task ends

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is the first follow-up of https://github.com/apache/spark/pull/21573 , which was only merged to 2.3.

This PR fixes the memory leak in another way: free the `UnsafeExternalMap` when the task ends. All the data buffers in Spark SQL are using `UnsafeExternalMap` and `UnsafeExternalSorter` under the hood, e.g. sort, aggregate, window, SMJ, etc. `UnsafeExternalSorter` registers a task completion listener to free the resource, we should apply the same thing to `UnsafeExternalMap`.

TODO in the next PR:
do not consume all the inputs when having limit in whole stage codegen.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @hvanhovell @viirya @gatorsmile

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -563,4 +563,12 @@ package object config {
       .intConf
       .checkValue(v => v > 0, "The value should be a positive integer.")
       .createWithDefault(2000)
+
+  private[spark] val BUFFER_PAGE_SIZE =
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: Small comment, while this is a valid change to make, it is a bit unrelated. Shall we do this in a different PR?

**Review by hvanhovell — APPROVED** [review]
LGTM - one minor nit.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
LGTM. `UnsafeExternalMap` in the description means `UnsafeFixedWidthAggregationMap` ?

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!
